File: Java/Tree/BJAlgo_5639_tree.java
import java.io.*;

public class BJAlgo_5639_tree {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        NodeMng bst = new NodeMng();

        String s = "";
        while ((s = br.readLine()) != null && s.length() != 0) {
            bst.add(Integer.parseInt(s));
        }

        bst.printValue(bst.head);
    }
}

class Node {
    int value;
    Node left;
    Node right;

    Node(int value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
}

class NodeMng {
    Node head;

    NodeMng() {
        this.head = null;
    }

    void add(int value) {
        if (this.head == null) {
            this.head = new Node(value);
            return;
        }

        Node currentNode = this.head;
        while (currentNode != null) {
            if (currentNode.value > value) {
                if (currentNode.left != null) {
                    currentNode = currentNode.left;
                } else {
                    currentNode.left = new Node(value);
                    break;
                }
            } else {
                if (currentNode.right != null) {
                    currentNode = currentNode.right;
                } else {
                    currentNode.right = new Node(value);
                    break;
                }
            }
        }
    }

    void printValue(Node node) {
        if (node.left != null) {
            printValue(node.left);
        }

        if (node.right != null) {
            printValue(node.right);
        }

        System.out.println(node.value);
    }
}

File: Python/Tree/BJAlgo_5639_tree.py
# 시간 초과

from sys import stdin

class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


class NodeMng:
    def __init__(self):
        self.head = None

    def add(self, value):
        if self.head is None:
            self.head = Node(value)
            return

        self.current_node = self.head
        while self.current_node is not None:
            if self.current_node.value > value:
                if self.current_node.left is not None:
                    self.current_node = self.current_node.left
                else:
                    self.current_node.left = Node(value)
                    break
            elif self.current_node.value < value:
                if self.current_node.right is not None:
                    self.current_node = self.current_node.right
                else:
                    self.current_node.right = Node(value)
                    break

    def printValue(self, node):
        if node.left is not None:
            self.printValue(node.left)

        if node.right is not None:
            self.printValue(node.right)

        print(node.value)


BST = NodeMng()
count = 0
while count <= 10000:
    try:
        num = int(stdin.readline())
    except:
        break

    BST.add(num)
    count += 1

BST.printValue(BST.head)
